api/src/wfl/module/copyfile.clj

```diff
@@ -14,57 +15,68 @@
   {:release "copyfile-v1.0"
    :top     "wdl/copyfile.wdl"})
 
-(defn- submit-workflow
+(defn ^:private get-cromwell-environment [{:keys [cromwell]}]
+  (let [envs (all/cromwell-environments #{:gotc-dev :gotc-prod} cromwell)]
+    (when (not= 1 (count envs))
+      (throw (ex-info "no unique environment matching Cromwell URL."
+               {:cromwell     cromwell
+                :environments envs})))
+    (first envs)))
+
+(defn ^:private submit-workflow
   "Submit WORKFLOW to Cromwell in ENVIRONMENT with OPTIONS and LABELS."
-  [environment workflow options labels]
+  [environment inputs options labels]
   (cromwell/submit-workflow
     environment
     (util/extract-resource (:top workflow-wdl))
     nil
-    (-> workflow (select-keys [:src :dst]) (util/prefix-keys pipeline))
+    (-> inputs (util/prefix-keys pipeline))
     options
     labels))
 
-(defn add-copyfile-workload!
-  "Use transaction TX to add the workload described by WORKLOAD-REQUEST."
-  [tx {:keys [items] :as _workload-request}]
-  (let [workflow-options (-> (:cromwell _workload-request)
-                             all/cromwell-environments
-                             first
-                             util/make-options
-                             (util/deep-merge (:workflow_options _workload-request)))
-        [id table]       (all/add-workload-table! tx workflow-wdl _workload-request)
-        to-row (fn [item] (assoc (:inputs item)
-                            :workflow_options
-                            (json/write-str (util/deep-merge workflow-options (:workflow_options item)))))]
-    (letfn [(add-id [m id] (assoc m :id id))]
-      (jdbc/insert-multi! tx table (map add-id (map to-row items) (range))))
-    id))
+(defn create-copyfile-workload!
+  "Use transaction TX to add the workload described by REQUEST."
+  [tx {:keys [items common] :as request}]
+  (letfn [(merge-to-json [shared specific]
+            (json/write-str (util/deep-merge shared specific)))
+          (serialize [workflow id]
+            (-> workflow
+              (assoc :id id)
+              (update :inputs #(merge-to-json (:inputs common) %))
+              (update :options #(merge-to-json (:options common) %))))]
+    (let [[id table] (batch/add-workload-table! tx workflow-wdl request)]
+      (jdbc/insert-multi! tx table (map serialize items (range)))
+      (workloads/load-workload-for-id tx id))))
```

Comment:
The `create-workload!` implementations for copyfile, wgs and xx are almost identical. In a subsequent change, we could explore reducing this technical debt. This means (in theory) we need to test only once for the "batch" workflows (or instantiate a test suite).